src/ccache.c

```diff
@@ -0,0 +1,523 @@
+	__asm__ __volatile__ goto(
+	    /* aw == Allocatable and writable section */
+	    ".pushsection __rseq_table, \"aw\"\n\t"
+	    ".balign 32\n\t"
+	    "cs_obj:\n\t"
+	    /* rseq_cs.version rseq_cs.flags */
+	    ".long 0, 0\n\t"
+	    /* rseq_cs.start_ip rseq_cs.post_commit_offset rseq_cs.abort_ip */
+	    ".quad rseq_cs_start%=, rseq_cs_postcommit%= - rseq_cs_start%=, rseq_cs_abort%=\n\t"
+	    ".popsection\n\t"
+
+	    "leaq cs_obj(%%rip), %%rax\n\t"
+	    "movq %%rax, %[rseq_cs]\n\t"
+
+	    /* -------------------- Begin ------------------- */
+	    "rseq_cs_start%=:\n\t"
+	    "cmpl %[cpu_id], %[initial_cpu_id]\n\t"
+	    /* Restart in case of CPU id missmatch*/
+	    "jne %l[restart]\n\t"
+
+	    /* rax = head */
+	    "movq %c[head_offset](%[bin]), %%rax\n\t"
+	    /* If in fallback mode, then jump to fallback label */
+	    "test %%rax, %%rax\n\t"
+	    "je %l[fallback]\n\t"
+	    /* If the bin is empty, then put into fallback mode and refill */
+	    "cmpq (%%rax), %%rax\n\t"
+	    "jne refill_not_needed%=\n\t"
+	    /* Prepare head = 0 */
+	    "xor %%rax, %%rax\n\t"
+	    "jmp commit%=\n\t"
+
+	    "refill_not_needed%=:\n\t"
+	    /* Store allocation in ret and preapare head++ */
+	    "movq (%%rax), %[ret]\n\t"
+	    "addq $8, %%rax\n\t"
+	    /* Commit head++ or head = 0 and exit CS */
+	    "commit%=:\n\t"
+	    "movq %%rax, %c[head_offset](%[bin])\n\t"
+
+	    /* ------------------- End ------------------- */
+	    "rseq_cs_postcommit%=:\n\t"
+
+	    /* ax == Allocatable and executable section */
+	    ".pushsection __rseq_abort, \"ax\"\n\t"
+	    ".byte 0x0f, 0x1f, 0x05\n\t"
+	    ".long "STRINGIFY(JEMALLOC_RSEQ_SIG)"\n\t"
+	    "rseq_cs_abort%=:\n\t"
+	    /*
+	     * Spin in case of abort - re-fetch per-cpu cache for the updated
+	     * CPU value in rseq_abi and retry.
+	     */
+	    "jmp rseq_cs_start%=\n\t"
+	    ".popsection\n\t"
+
+	    : [ret] "=&r" (ret)
+	    : [bin] "r" (bin),
+	      [head_offset] "i" (offsetof(ccache_bin_t, head)),
+	      [rseq_cs] "m" (rseq_abi->rseq_cs),
+	      [cpu_id] "m" (rseq_abi->cpu_id_start),
+	      [initial_cpu_id] "r" (cpu),
+	      [ccache] "rm" (ccache)
+	    : "rax", "cc"
+	    : restart, fallback
+	);
+	if (unlikely(bin->head == NULL)) {
+		/*
+		 * At this point, we're in the refill/flush mode. This thread is
+		 * the owner of the refill, other threads willing to draw values
+		 * from 'cpu' will jump to fallback label and return NULL.
+		 */
+		if (small) {
+			ret =
+			    ccache_bin_refill(tsd, ccache, bin, arena,
+				ind, small);
+		} else {
+			/*
+			 * Do not refill if large size class; transfer to the 
+			 * empty state. Arena will handle zero and stats.
+			 */
+			bin->head = (void **)&bin->head;
+fallback:
+			return arena_malloc_hard(tsd_tsdn(tsd), arena, size,
+			    ind, zero);
+		}
+	}
+	assert(ret != NULL);
+	if (unlikely(zero)) {
+		size_t usize = sz_index2size(ind);
+		assert(tcache_salloc(tsd_tsdn(tsd), ret) == usize);
+		memset(ret, 0, usize);
```

Comment:
Will want to monitor the cost of this part -- since we are targeting "medium" size classes, the zero pattern there might be different than the small ones. Imagine a recycling pattern which goes through `ccache_bin_refill` often to grab 4K-14K sizes. The refill may grab many fresh zeroed pages, but the zero info is lost when stored in ccache.

Reply:
That is a nice observation, I'll keep an eye on it.